### ISSUE TITLE
When reading JSON, Error in procedure read-control-char: In procedure integer->char: Argument 1 out of range: 55357

### ISSUE THREAD

**Issue by pkillnine** [aconchillo/guile-json#58]
When trying to read the JSON at "https://www.reddit.com/r/programming/comments/.json?count=25&after=t1_fyuzqiw", this error is returned:

```
Backtrace:
In ice-9/boot-9.scm:
  1736:10 18 (with-exception-handler _ _ #:unwind? _ # _)
In unknown file:
          17 (apply-smob/0 #<thunk 7fb403483e60>)
In ice-9/boot-9.scm:
    718:2 16 (call-with-prompt _ _ #<procedure default-prompt-handle…>)
In ice-9/eval.scm:
    619:8 15 (_ #(#(#<directory (guile-user) 7fb40307bf00>)))
In ice-9/boot-9.scm:
   2806:4 14 (save-module-excursion _)
  4351:12 13 (_)
In /home/itsme/guile-read-json.scm:
    34:12 12 (get-all-authors #:subreddit _ #:after _)
In json/parser.scm:
   327:15 11 (json->scm _ #:null _)
   177:20 10 (json-read-object #<input: string 7fb40162f5b0> null)
   159:18  9 (read-pair #<input: string 7fb40162f5b0> null)
   177:20  8 (json-read-object #<input: string 7fb40162f5b0> null)
   159:18  7 (read-pair #<input: string 7fb40162f5b0> null)
   214:21  6 (json-read-array #<input: string 7fb40162f5b0> null)
   177:20  5 (json-read-object #<input: string 7fb40162f5b0> null)
   159:18  4 (read-pair #<input: string 7fb40162f5b0> null)
   177:20  3 (json-read-object #<input: string 7fb40162f5b0> null)
   159:18  2 (read-pair #<input: string 7fb40162f5b0> null)
   279:18  1 (json-read-string #<input: string 7fb40162f5b0>)
    249:2  0 (read-control-char _)

json/parser.scm:249:2: In procedure read-control-char:
In procedure integer->char: Argument 1 out of range: 55357
```

To reproduce this, run this in Guile:

```
(use-modules (web client) (web response) (json) (rnrs bytevectors))

(define* (get-reddit-response)
  (let ((uri "https://www.reddit.com/r/programming/comments/.json?count=25&after=t1_fyuzqiw"))
  
  (json-string->scm
   (utf8->string
    (u8-list->bytevector
     (u8vector->list
      (read-response-body (http-request uri #:streaming? #t))))))))

(get-reddit-response)
```

I asked about this in the #guile channel when I tried with another page of Reddit JSON comments, there was some discussion about the cause of the bug and how to fix it, but I don't know Guile well enough to fix it myself:

```
<pkill9> how can I deal with unsupported characters in text that gets passed to guile-json
<pkill9> i want to just replace all non-supported characters in a string basically
<pkill9> the json parser passes some text to integer->char, and then an error: json/parser.scm:249:2: In procedure read-control-char:
<pkill9> In procedure integer->char: Argument 1 out of range:
<pkill9> 55358
<pkill9> 55358 i believe is a hugging emoji, lol
<RhodiumToad> no its not
<RhodiumToad> 55358 = U+D83E, which is a high surrogate, and therefore not a valid character
<RhodiumToad> I believe json encodes surrogate pairs separately? so you might have to fetch both parts of the pair and convert them together
<RhodiumToad> U+D83E could be the first half of U+1F917, which would encode as U+D83E U+DD17
<RhodiumToad> i.e. "\uD83E\uDD17"  in json
<RhodiumToad> (U+1F917 == hugging face emoji)
<dsmith-work> So that's a bug in guile-json?
<RhodiumToad> if it's trying to decode \uXXXX independently when the value is a surrogate, yes
<dsmith-work> Ugh.
<RhodiumToad> (it's kind of ick that json uses that encoding in the first place...)
<RhodiumToad> haha. "strictly complies to http://json.org specification".
<RhodiumToad> anyway, yes, definite bug in guile-json.
<pkill9> guix's guile-json is 3 stable versions out of date, so it may have been fixed, i'll see
<RhodiumToad> I was looking at what was on savannah
<RhodiumToad> http://git.savannah.nongnu.org/cgit/guile-json.git/tree/json/parser.scm#n275
<holomorph> what happened to ice-9 json
<RhodiumToad> it is not correct that "Characters in Guile match the JSON representation" so that breaks
<RhodiumToad> I guess the fix should be to read 4 hex digits, and if the value is in DC00-DFFF throw error, if it's D800-DBFF then read in an immediately following \uXXXX and error if that isn't DC00-DFFF
```

**Comment by aconchillo** [comment]
Thank you for reporting this issue. I have been able to reproduce it with latest version and fixing the issue.

**Comment by pkillnine** [comment]
RhodiumToad from the #Guile channel posted a fix which I tested and works when you replace `(define (read-unicode-char ...` in `json/parser.scm` with this:

```
(define (read-hex-value port)
  (+ (* 4096 (read-hex-digit->integer port))
     (* 256 (read-hex-digit->integer port))
     (* 16 (read-hex-digit->integer port))
     (read-hex-digit->integer port)))

;; The initial \u has been read, so we're looking at either:
;;   XXXX  (where the value is <D800 or >DFFF)
;; or
;;   XXXX\uXXXX   if the first XXXX is in D800..DBFF
;; (a value of DC00..DFFF as the first of a pair is invalid)

(define (read-unicode-char port)
  (let ((codepoint (read-hex-value port)))
    (cond
     ((and (>= codepoint #xD800) (< codepoint #xDC00))
      (expect-delimiter port #\\)
      (expect-delimiter port #\u)
      (let ((low-surrogate (read-hex-value port)))
        (when (or (>= low-surrogate #xE000)
                  (< low-surrogate #xDC00))
          (json-exception port))
        ;; final codepoint is the (high surrogate - D800)<<10,
        ;; plus the low surrogate - DC00, plus 10000,
        ;; which is the same as (hi<<10)+lo - 35FDC00
        (integer->char (+ (* codepoint #x400) low-surrogate #x-35FDC00))))
     ((and (>= codepoint #xDC00) (< codepoint #xE000))
      (json-exception port))
     (else
      (integer->char codepoint)))))
```

**Comment by aconchillo** [comment]
Oh, thanks! I actually fixed it in a very similar way. That surrogate pair generated a thumbs up character, I have to double-check. Had to went outside but I will release a fix later today  Thanks again!

**Comment by aconchillo** [comment]
@pkillnine Just released guile-json 4.3.1 which fixes this. I added a note to thank you both on the NEWS file.

**Comment by pkillnine** [comment]
Apparently there is a bug with the fix, RhodiumToad said in #guile:

```
<RhodiumToad> pkill9: looks like it's now broken for \uE000 and upwards?
<RhodiumToad> my version handled that ok, but the committed one doesn't
```

**Comment by aconchillo** [comment]
Ooops! You are totally right! Good catch.

**Comment by aconchillo** [comment]
I think it should be fixed now. I just released 4.3.2. I'm so sorry about that. Didn't pay enough attention yesterday. Thank you!
